### ISSUE TITLE
Question - ServiceEndpoint Rest Client example

### ISSUE THREAD

**Issue by jsalado** [Microsoft/azure-devops-extension-api#12]
Hello,

I'm trying to use the ServiceEndpointRestClient and I got stuck. I couldn't find documentation out there... 

Can you provide any examples?

What I want to do is call one of the data sources defined in a custom endpoint I have defined in my extension using executeServiceEndpointRequest().

This is all done from a web hub contribution so I can show the data returned by my endpoint in the html associated to the web hub when I click it. Does this make sense at all?

Thanks,

**Comment by mkaszub** [comment]
Hello,

I need it as well. @jsalado any luck on calling it?

Thanks,
Marcin

**Comment by jsalado** [comment]
Hello @mkaszub,

no, no luck. I was waiting for someone to help out here, but since there was no answer I had to re invent the wheel and code the REST call to the endpoint myself using node httpsRequest

I hope someone from Microsoft Azure DevOps reads this and can tell us at least if this is possible or going to be possible

**Comment by mkaszub** [comment]
@jsalado 

Yeah Microsoft is helpful as always, new API every 2 months without documentation... :( its support for developers is sometimes so bad ehhhh...


Anyway! I made it. This is the peace of code that should work for you as well. Probably there are some places for improvements...

So first what you have to do is get an endpoint you want to use


```
const projectService = await SDK.getService<IProjectPageService>(CommonServiceIds.ProjectPageService);
let project = await projectService.getProject();
    
let client = await getClient(ServiceEndpointRestClient)
let endpoints = await this._client.getServiceEndpoints(project.id, "some_type_of_endpoint")
let endpoint = endpoints[0]
```

Then you have to create ServiceEndpointRequest that contains all information about your request

```
let resource = "/example/path"
let content = "if post probably not empty"
let type = "POST"
const ds: DataSourceDetails = {
    dataSourceName: null,
    dataSourceUrl: "{{endpoint.url}}/" + resource,
    headers: [],
    initialContextTemplate: null,
    parameters: {},
    requestContent: content,
    requestVerb: type,
    resourceUrl: null,
    resultSelector: "jsonpath:$",
}
const request: ServiceEndpointRequest = {
    dataSourceDetails: ds,
    resultTransformationDetails: null,
    serviceEndpointDetails: null
 }
```

And then send request through endpoint

`let response = await client.executeServiceEndpointRequest(request, project.name,  endpoint.id)`

Done!

**Comment by jsalado** [comment]
@mkaszub 

Looks great! thanks a lot for sharing. I'll try it on my extension as soon as I can and let you know how it worked for me.

Let's see if Microsoft pick up the ball now ;-)

Cheers,
J.

**Comment by shajusukumaran** [comment]
@jsalado @mkaszub i am trying to implement something similar and used the sample above as reference to connect to an azure function api from my custom extension, but i am getting 401 error (`"Failed to query service connection API: 'https://xxxx.azurewebsites.net/api/xxxxx'. Status Code: 'Unauthorized', Response from server: ''"`). The same service connection endpoint works perfectly from within a task. Any idea why it would be giving an error when called using `await client.executeServiceEndpointRequest(request, projectName, endpoint.id);`?

I am using token based authorization.

**Comment by Dejulia489** [comment]
> @jsalado
> 
> Yeah Microsoft is helpful as always, new API every 2 months without documentation... :( its support for developers is sometimes so bad ehhhh...
> 
> Anyway! I made it. This is the peace of code that should work for you as well. Probably there are some places for improvements...
> 
> So first what you have to do is get an endpoint you want to use
> 
> ```
> const projectService = await SDK.getService<IProjectPageService>(CommonServiceIds.ProjectPageService);
> let project = await projectService.getProject();
>     
> let client = await getClient(ServiceEndpointRestClient)
> let endpoints = await client.getServiceEndpoints(project.id, "some_type_of_endpoint")
> let endpoint = endpoints[0]
> ```
> 
> Then you have to create ServiceEndpointRequest that contains all information about your request
> 
> ```
> let resource = "/example/path"
> let content = "if post probably not empty"
> let type = "POST"
> const ds: DataSourceDetails = {
>     dataSourceName: null,
>     dataSourceUrl: "{{endpoint.url}}/" + resource,
>     headers: [],
>     initialContextTemplate: null,
>     parameters: {},
>     requestContent: content,
>     requestVerb: type,
>     resourceUrl: null,
>     resultSelector: "jsonpath:$",
> }
> const request: ServiceEndpointRequest = {
>     dataSourceDetails: ds,
>     resultTransformationDetails: null,
>     serviceEndpointDetails: null
>  }
> ```
> 
> And then send request through endpoint
> 
> `let response = await client.executeServiceEndpointRequest(request, project.name, endpoint.id)`
> 
> Done!
> 
> BTW
> If you do not want to use url as resource just provide resource name defined in your plugin


How is ServiceEndpointRestClient defined?

**Comment by LockTar** [comment]
Hi, Can someone give more information how to call an external API?
I just want to call an external api with in example basic or token auth. 
The documentation is terrible. 

I also asked this on [StackOverflow](https://stackoverflow.com/questions/67221757/create-custom-azure-devops-service-connection-for-rest-api) but without any success.
